### DISCUSSION TITLE
should I link Jemalloc ?

### DISCUSSION THREAD

**Discussion by foTok** [duckdb/duckdb#5608] [Q&A]
v0.6 introduced jemalloc memory allocator in Linux. I noticed that the source code of jemalloc is included in v0.6, however, when I compile my (C++)program, I still have to link jemalloc (-ljemalloc), or segment fault happens when checkpoint.

**Comment by hannes** [comment]
Can you please try to create a minimal reproducible example? the linking should not be required.

**Reply by foTok** [reply]
The problem occurs in a complex situation, I am try to reproduce it in a minimal case.
By now, there are two hints.

1. I compiled duckdb from source by CLion, and got a dynamic library libduckdb.so in duckdb/cmake-build-release/src with size of 26399KB, however, the official released libduckdb.so is 40553KB. The lib sizes are quite different.

2. When crash at checkpoint (not every checkpoint crash), the following shows in the terminal:
seems something wrong about FSSTStorage.
-bash-4.2$ *** Error in `./DuckDBVerification': free(): invalid next size (normal): 0x0000000004d1e510 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x6dce1)[0x7fdc44eeace1]
/lib64/libc.so.6(+0x734c6)[0x7fdc44ef04c6]
/lib64/libc.so.6(+0x74213)[0x7fdc44ef1213]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(+0x118dc77)[0x7fdc46ee6c77]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb11FSSTStorage18StringFinalAnalyzeERNS_12AnalyzeStateE+0x11f)[0x7fdc46e4b3ef]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb22ColumnDataCheckpointer27DetectBestCompressionMethodERm+0x170)[0x7fdc46e9fbd0]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb22ColumnDataCheckpointer11WriteToDiskEv+0x92)[0x7fdc46e9ffd2]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb10ColumnData10CheckpointERNS_8RowGroupERNS_19PartialBlockManagerERNS_20ColumnCheckpointInfoE+0x142)[0x7fdc46ea38d2]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18StandardColumnData10CheckpointERNS_8RowGroupERNS_19PartialBlockManagerERNS_20ColumnCheckpointInfoE+0x40)[0x7fdc46ea3d90]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb8RowGroup11WriteToDiskERNS_19PartialBlockManagerERKSt6vectorINS_15CompressionTypeESaIS4_EE+0xb7)[0x7fdc46ea4857]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb8RowGroup10CheckpointERNS_14RowGroupWriterERSt6vectorISt10unique_ptrINS_14BaseStatisticsESt14default_deleteIS5_EESaIS8_EE+0x86)[0x7fdc46ea4b06]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18RowGroupCollection10CheckpointERNS_15TableDataWriterERSt6vectorISt10unique_ptrINS_14BaseStatisticsESt14default_deleteIS5_EESaIS8_EE+0x5d)[0x7fdc46ea508d]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb9DataTable10CheckpointERNS_15TableDataWriterE+0xc7)[0x7fdc46ec8d17]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb16CheckpointWriter10WriteTableERNS_17TableCatalogEntryE+0x8a)[0x7fdc46ebb04a]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb16CheckpointWriter11WriteSchemaERNS_18SchemaCatalogEntryE+0x5b5)[0x7fdc46ebb675]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb26SingleFileCheckpointWriter16CreateCheckpointEv+0x1a7)[0x7fdc46ec7f47]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb24SingleFileStorageManager16CreateCheckpointEbb+0xa0)[0x7fdc46ec81d0]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18TransactionManager17CommitTransactionB5cxx11ERNS_13ClientContextEPNS_11TransactionE+0x346)[0x7fdc46edf546]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18TransactionContext6CommitEv+0x42)[0x7fdc46edf732]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext16EndQueryInternalERNS_17ClientContextLockEbb+0x1b9)[0x7fdc46dfef19]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext15CleanupInternalERNS_17ClientContextLockEPNS_15BaseQueryResultEb+0x7f)[0x7fdc46dff20f]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext13FetchInternalERNS_17ClientContextLockERNS_8ExecutorERNS_15BaseQueryResultE+0x4e)[0x7fdc46dff79e]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext19FetchResultInternalERNS_17ClientContextLockERNS_18PendingQueryResultE+0x298)[0x7fdc46e01e58]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18PendingQueryResult15ExecuteInternalERNS_17ClientContextLockE+0x72)[0x7fdc46e02582]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext27ExecutePendingQueryInternalERNS_17ClientContextLockERNS_18PendingQueryResultE+0xd)[0x7fdc46e025bd]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext5QueryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEb+0x228)[0x7fdc46e04618]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb10Connection5QueryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x18)[0x7fdc46e04958]
./DuckDBVerification[0x43cf67]
./DuckDBVerification[0x40eac0]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fdc44e9d835]
./DuckDBVerification[0x40f849]

**Reply by lnkuiper** [reply]
Could you try to build DuckDB from the command line instead and see if the problem is still there?

Also, when compiling through CLion,be sure to set `-DBUILD_JEMALLOC_EXTENSION=1` under CMake options.

**Reply by foTok** [reply]
> Could you try to build DuckDB from the command line instead and see if the problem is still there?
> 
> Also, when compiling through CLion,be sure to set `-DBUILD_JEMALLOC_EXTENSION=1` under CMake options.

I tried the official v0.6.0 release and did not link jemalloc. The checkpoint crash still exits.
./DuckDBVerification': free(): invalid next size (normal): 0x00007f65b1d9fce0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x6dce1)[0x7f663c920ce1]
/lib64/libc.so.6(+0x734c6)[0x7f663c9264c6]
/lib64/libc.so.6(+0x74213)[0x7f663c927213]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(+0x13560b5)[0x7f663eae50b5]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb11FSSTStorage18StringFinalAnalyzeERNS_12AnalyzeStateE+0x182)[0x7f663ea4e032]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb22ColumnDataCheckpointer27DetectBestCompressionMethodERm+0x18c)[0x7f663eaa088c]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb22ColumnDataCheckpointer11WriteToDiskEv+0xa8)[0x7f663eaa0cd8]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb10ColumnData10CheckpointERNS_8RowGroupERNS_19PartialBlockManagerERNS_20ColumnCheckpointInfoE+0x160)[0x7f663eaa1180]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18StandardColumnData10CheckpointERNS_8RowGroupERNS_19PartialBlockManagerERNS_20ColumnCheckpointInfoE+0x4e)[0x7f663eaa148e]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb8RowGroup11WriteToDiskERNS_19PartialBlockManagerERKSt6vectorINS_15CompressionTypeESaIS4_EE+0x123)[0x7f663eaa2af3]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb8RowGroup10CheckpointERNS_14RowGroupWriterERSt6vectorISt10unique_ptrINS_14BaseStatisticsESt14default_deleteIS5_EESaIS8_EE+0xbe)[0x7f663eaa2cde]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18RowGroupCollection10CheckpointERNS_15TableDataWriterERSt6vectorISt10unique_ptrINS_14BaseStatisticsESt14default_deleteIS5_EESaIS8_EE+0x75)[0x7f663eaa3235]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb9DataTable10CheckpointERNS_15TableDataWriterE+0x226)[0x7f663eabd476]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb16CheckpointWriter10WriteTableERNS_17TableCatalogEntryE+0x98)[0x7f663eab68f8]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb16CheckpointWriter11WriteSchemaERNS_18SchemaCatalogEntryE+0x5fd)[0x7f663eab730d]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb26SingleFileCheckpointWriter16CreateCheckpointEv+0x1e7)[0x7f663eac24d7]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb24SingleFileStorageManager16CreateCheckpointEbb+0x93)[0x7f663eac2813]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18TransactionManager17CommitTransactionB5cxx11ERNS_13ClientContextEPNS_11TransactionE+0x484)[0x7f663eadbcc4]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18TransactionContext6CommitEv+0x51)[0x7f663eadbe11]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext16EndQueryInternalERNS_17ClientContextLockEbb+0x1c9)[0x7f663ea03c99]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext15CleanupInternalERNS_17ClientContextLockEPNS_15BaseQueryResultEb+0x98)[0x7f663ea04148]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext13FetchInternalERNS_17ClientContextLockERNS_8ExecutorERNS_15BaseQueryResultE+0x90)[0x7f663ea04560]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext19FetchResultInternalERNS_17ClientContextLockERNS_18PendingQueryResultE+0x274)[0x7f663ea04bf4]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb18PendingQueryResult15ExecuteInternalERNS_17ClientContextLockE+0x60)[0x7f663ea05920]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext27ExecutePendingQueryInternalERNS_17ClientContextLockERNS_18PendingQueryResultE+0x20)[0x7f663ea059b0]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb13ClientContext5QueryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEb+0x258)[0x7f663ea0d1d8]
/home/duck/workspace/DuckDBVerification/lib/libduckdb.so(_ZN6duckdb10Connection5QueryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x25)[0x7f663ea0d665]
./DuckDBVerification[0x44a3c6]
/home/duck/software/gcc-11.2.0/lib64/libstdc++.so.6(+0xd1340)[0x7f663d23d340]
/lib64/libpthread.so.0(+0x86e4)[0x7f663d57a6e4]
/lib64/libc.so.6(clone+0x6d)[0x7f663c9992dd]

**Reply by foTok** [reply]
> Could you try to build DuckDB from the command line instead and see if the problem is still there?
> 
> Also, when compiling through CLion,be sure to set `-DBUILD_JEMALLOC_EXTENSION=1` under CMake options.

I think there is something wrong with the *FSST compression functionality*.
I commented some code in *compression_config.cpp* as follows, and then no crash happened again for over 10 hours.
```
static DefaultCompressionMethod internal_compression_methods[] = {
    /* some omitted code */
//    {CompressionType::COMPRESSION_FSST, FSSTFun::GetFunction, FSSTFun::TypeIsSupported}, <-- commented
    {CompressionType::COMPRESSION_AUTO, nullptr, nullptr}};

vector<CompressionFunction *> DBConfig::GetCompressionFunctions(PhysicalType data_type) {
	/* some omitted code */
//	TryLoadCompression(*this, result, CompressionType::COMPRESSION_FSST, data_type); <-- commented
	return result;
}
```
